File: July-LeetCoding-Challenge-2020/Week3/79-word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 * Runtime: 88 ms, faster than 91.28% of JavaScript online submissions for Word Search.
 * Memory Usage: 38.6 MB, less than 72.31% of JavaScript online submissions for Word Search.
 * TC: O(N^2), where N is number of elements in a board
 * SC: O(N), where N is number of elements in a board
 */
const vDir = [0, 0, -1, 1];
const hDir = [-1, 1, 0, 0];

var exist = function(board, word) {
    if (!board || !board[0] || board.length === 0 || board[0].length === 0) {
        return false;
    } else if (!word || word.length === 0) {
        return true;
    }

    for (let i = 0; i < board.length; i++) {
        for (let j = 0; j < board[0].length; j++) {
            if (traverse(board, word, i, j, 0)) {
                return true;
            }
        }
    }

    return false;
};

function traverse (board, word, i, j, depth) {
    if (depth === word.length) {
        return true;
    } else if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) {
        return false;
    }

    const char = board[i][j];

    if (char === "#" || char !== word.charAt(depth)) {
        return false;
    }

    board[i][j] = "#";

    for (let dir = 0; dir <= 3; dir++) {
        if (traverse(board, word, i + hDir[dir], j + vDir[dir], depth + 1)) {
            return true;
        }
    }

    board[i][j] = char;

    return false;
}

File: July-LeetCoding-Challenge-2020/Week3/79-Word-Search.cpp
// Solution #1 Depth First Search
// TC: O(# of row * # of col * 4^(length of word))
// SC: O(length of word)
// Runtime: 412 ms, faster than 21.52%
// Memory Usage: 43.1 MB, less than 45.66%
class Solution {
    vector<int> row{0,1,0,-1}, col{1,0,-1,0};
    map<pair<int,int>, bool> visited;
public:
    
    bool exist(vector<vector<char>>& board, string word) {
                
        for(int i = 0; i < board.size(); i++)
            for(int j = 0; j < board[0].size(); j++)
                if(board[i][j] == word[0]){
                    if(wordExists(board, word, i, j, 0))
                        return true;
                    visited.clear();
                }
                    
        return false;
    }
    
    bool wordExists(vector<vector<char>> &board, string word, int i, int j, int index){
        if(index >= word.length())
            return true;
        if(!isInBound(board,i,j) || !isCharMatch(board[i][j],word[index]) || visited[{i,j}])
            return false;
        
        visited[{i,j}] = true;
        bool res = false;
        
        for(int k = 0; k < 4; k++){
            res = wordExists(board, word, i+row[k], j+col[k], index+1);
            if(res)
                break;
        }
        
        visited[{i,j}] = false;
        return res;
    }
    
    bool isInBound(vector<vector<char>> &board, int i, int j){
        return i >= 0 && i < board.size() && j >= 0 && j < board[0].size();
    }
    
    bool isCharMatch(char c, char w){
        return c == w;
    }
};